DataDog/dd-trace-rb
# Add support for `http.client_ip` tag for Rack-based frameworks

**What does this PR do?**


Add support for the `http.client_ip` tag for Rack-based API frameworks.

**Motivation**


The `http.client_ip` tag will help distinguish whether the request came from a public or a private address, allowing Datadog to distinguish external from internal services and API endpoints.

**How to test the change?**


The change is validated by RSpec behavior tests.

To test integration:

Set up and start a demo Rails server and enable auto-instrumentation.
Then use the following commands:

```bash
curl localhost:3000 # direct IP
curl -H 'X-Forwarded-For: 1.1.1.1' localhost:3000 # forwarded IP
curl -H 'X-Forwarded-For: 1.1.1.1' -H 'X-Forwarded-For: 1.1.1.1' localhost:3000 # two values for the header = no client ip
curl -H 'X-Forwarded-For: 1.1.1.1' -H 'X-Real-Ip: 1.2.3.4' localhost:3000 # two "ip" headers set = no client ip
```

Review the logs from the server. The emitted traces should reflect the expected results.
